File: Sources/QuotaGlanceCore/Providers/MiniMaxModelRemainsStrategy.swift
import Foundation

/// Named parse strategy `miniMaxModelRemains` (Contracts/README.md, "Named
/// parse strategies"): derives quota windows from MiniMax `model_remains`
/// entries. Direct windows win over interval/weekly windows, counted forms
/// win over percent forms, an absent status counts as active (`!= 0`), and
/// timestamps are epoch seconds or milliseconds (magnitude > 1e10 ⇒
/// milliseconds).
enum MiniMaxModelRemainsStrategy {
    static func quotaWindows(from value: SpecJSONValue?) -> [QuotaWindow] {
        guard case let .array(entries) = value else { return [] }
        return entries.flatMap(quotaWindows(for:))
    }

    static func quotaWindows(for entry: SpecJSONValue) -> [QuotaWindow] {
        if let direct = directQuotaWindow(for: entry) {
            return [direct]
        }

        var windows: [QuotaWindow] = []
        if decimal(entry, "current_interval_status") != 0,
           let interval = intervalQuotaWindow(for: entry) {
            windows.append(interval)
        }
        if decimal(entry, "current_weekly_status") != 0,
           let weekly = weeklyQuotaWindow(for: entry) {
            windows.append(weekly)
        }
        return windows
    }

    private static func directQuotaWindow(for entry: SpecJSONValue) -> QuotaWindow? {
        let limit = decimal(entry, "total")
        let suppliedUsed = decimal(entry, "used")
        let suppliedRemaining = decimal(entry, "remains")
        guard limit != nil || suppliedUsed != nil || suppliedRemaining != nil else {
            return nil
        }

        let used = suppliedUsed ?? subtract(limit, suppliedRemaining)
        let remaining = suppliedRemaining ?? subtract(limit, suppliedUsed)
        return QuotaWindow(
            label: normalizedLabel(string(entry, "label"))
                ?? "\(normalizedModelName(string(entry, "model_name"))) quota",
            used: used,
            limit: limit,
            remaining: remaining,
            unit: normalizedLabel(string(entry, "unit")) ?? "requests",
            resetsAt: providerDate(decimal(entry, "reset_time") ?? decimal(entry, "end_time"))
        )
    }

    private static func intervalQuotaWindow(for entry: SpecJSONValue) -> QuotaWindow? {
        let label = isFiveHourWindow(
            start: decimal(entry, "start_time"),
            end: decimal(entry, "end_time")
        )
            ? "\(normalizedModelName(string(entry, "model_name"))) 5-hour quota"
            : "\(normalizedModelName(string(entry, "model_name"))) quota"
        return countedOrPercentWindow(
            label: label,
            total: decimal(entry, "current_interval_total_count"),
            used: decimal(entry, "current_interval_usage_count"),
            remainingPercent: decimal(entry, "current_interval_remaining_percent"),
            resetsAt: decimal(entry, "end_time")
        )
    }

    private static func weeklyQuotaWindow(for entry: SpecJSONValue) -> QuotaWindow? {
        countedOrPercentWindow(
            label: "\(normalizedModelName(string(entry, "model_name"))) weekly quota",
            total: decimal(entry, "current_weekly_total_count"),
            used: decimal(entry, "current_weekly_usage_count"),
            remainingPercent: decimal(entry, "current_weekly_remaining_percent"),
            resetsAt: decimal(entry, "weekly_end_time")
        )
    }

    private static func countedOrPercentWindow(
        label: String,
        total: Decimal?,
        used: Decimal?,
        remainingPercent: Decimal?,
        resetsAt: Decimal?
    ) -> QuotaWindow? {
        if total.map({ $0 > 0 }) == true || used.map({ $0 > 0 }) == true {
            return QuotaWindow(
                label: label,
                used: used,
                limit: total,
                remaining: subtract(total, used),
                unit: "requests",
                resetsAt: providerDate(resetsAt)
            )
        }
        if let percentage = remainingPercent {
            let remaining = max(0, min(100, percentage))
            return QuotaWindow(
                label: label,
                used: 100 - remaining,
                limit: 100,
                remaining: remaining,
                unit: "%",
                resetsAt: providerDate(resetsAt)
            )
        }
        return nil
    }

    private static func subtract(_ lhs: Decimal?, _ rhs: Decimal?) -> Decimal? {
        guard let lhs, let rhs else { return nil }
        return lhs - rhs
    }

    private static func normalizedLabel(_ value: String?) -> String? {
        guard let value else { return nil }
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }

    private static func normalizedModelName(_ value: String?) -> String {
        normalizedLabel(value) ?? "Token Plan"
    }

    private static func providerDate(_ value: Decimal?) -> Date? {
        guard let value else { return nil }
        var seconds = NSDecimalNumber(decimal: value).doubleValue
        if abs(seconds) > 10_000_000_000 {
            seconds /= 1_000
        }
        return Date(timeIntervalSince1970: seconds)
    }

    private static func isFiveHourWindow(start: Decimal?, end: Decimal?) -> Bool {
        guard let startDate = providerDate(start), let endDate = providerDate(end) else {
            return false
        }
        return abs(endDate.timeIntervalSince(startDate) - 18_000) < 1
    }

    private static func decimal(_ entry: SpecJSONValue, _ key: String) -> Decimal? {
        guard case let .object(fields) = entry else { return nil }
        return SpecEngine.decimal(fields[key])?.value
    }

    private static func string(_ entry: SpecJSONValue, _ key: String) -> String? {
        guard case let .object(fields) = entry,
              case let .string(value)? = fields[key] else {
            return nil
        }
        return value
    }
}

File: Platforms/Android/app/src/main/java/com/liangrui/quotaglance/core/MiniMaxModelRemainsStrategy.kt
package com.liangrui.quotaglance.core

import java.math.BigDecimal
import kotlin.math.abs
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

/**
 * Named parse strategy allowed by the provider contract for MiniMax token-plan
 * remains. It is deliberately shape-only: all provider policy remains in the
 * shared spec.
 */
internal object MiniMaxModelRemainsStrategy {
    fun windows(value: JsonElement?): List<QuotaWindow> =
        (value as? JsonArray).orEmpty().flatMap(::windowsFor)

    private fun windowsFor(entry: JsonElement): List<QuotaWindow> {
        directWindow(entry)?.let { return listOf(it) }
        return buildList {
            if (decimal(entry, "current_interval_status") != BigDecimal.ZERO) {
                intervalWindow(entry)?.let(::add)
            }
            if (decimal(entry, "current_weekly_status") != BigDecimal.ZERO) {
                weeklyWindow(entry)?.let(::add)
            }
        }
    }

    private fun directWindow(entry: JsonElement): QuotaWindow? {
        val limit = decimal(entry, "total")
        val suppliedUsed = decimal(entry, "used")
        val suppliedRemaining = decimal(entry, "remains")
        if (limit == null && suppliedUsed == null && suppliedRemaining == null) return null
        return QuotaWindow(
            label = normalizedLabel(string(entry, "label"))
                ?: "${normalizedModelName(string(entry, "model_name"))} quota",
            used = suppliedUsed ?: subtract(limit, suppliedRemaining),
            limit = limit,
            remaining = suppliedRemaining ?: subtract(limit, suppliedUsed),
            unit = normalizedLabel(string(entry, "unit")) ?: "requests",
            resetsAtMillis = providerDateMillis(decimal(entry, "reset_time") ?: decimal(entry, "end_time")),
        )
    }

    private fun intervalWindow(entry: JsonElement): QuotaWindow? {
        val label = if (isFiveHourWindow(decimal(entry, "start_time"), decimal(entry, "end_time"))) {
            "${normalizedModelName(string(entry, "model_name"))} 5-hour quota"
        } else {
            "${normalizedModelName(string(entry, "model_name"))} quota"
        }
        return countedOrPercentWindow(
            label = label,
            total = decimal(entry, "current_interval_total_count"),
            used = decimal(entry, "current_interval_usage_count"),
            remainingPercent = decimal(entry, "current_interval_remaining_percent"),
            resetsAt = decimal(entry, "end_time"),
        )
    }

    private fun weeklyWindow(entry: JsonElement): QuotaWindow? = countedOrPercentWindow(
        label = "${normalizedModelName(string(entry, "model_name"))} weekly quota",
        total = decimal(entry, "current_weekly_total_count"),
        used = decimal(entry, "current_weekly_usage_count"),
        remainingPercent = decimal(entry, "current_weekly_remaining_percent"),
        resetsAt = decimal(entry, "weekly_end_time"),
    )

    private fun countedOrPercentWindow(
        label: String,
        total: BigDecimal?,
        used: BigDecimal?,
        remainingPercent: BigDecimal?,
        resetsAt: BigDecimal?,
    ): QuotaWindow? {
        if (total?.compareTo(BigDecimal.ZERO) == 1 || used?.compareTo(BigDecimal.ZERO) == 1) {
            return QuotaWindow(
                label = label,
                used = used,
                limit = total,
                remaining = subtract(total, used),
                unit = "requests",
                resetsAtMillis = providerDateMillis(resetsAt),
            )
        }
        remainingPercent ?: return null
        val remaining = remainingPercent.coerceIn(BigDecimal.ZERO, BigDecimal(100))
        return QuotaWindow(
            label = label,
            used = BigDecimal(100).subtract(remaining),
            limit = BigDecimal(100),
            remaining = remaining,
            unit = "%",
            resetsAtMillis = providerDateMillis(resetsAt),
        )
    }

    private fun subtract(left: BigDecimal?, right: BigDecimal?): BigDecimal? =
        if (left == null || right == null) null else left.subtract(right)

    private fun normalizedLabel(value: String?): String? = value?.trim()?.takeIf(String::isNotEmpty)
    private fun normalizedModelName(value: String?): String = normalizedLabel(value) ?: "Token Plan"

    private fun providerDateMillis(value: BigDecimal?): Long? {
        value ?: return null
        return if (value.abs() > BigDecimal("10000000000")) exactLong(value)
        else exactLong(value.multiply(BigDecimal(1000)))
    }

    private fun isFiveHourWindow(start: BigDecimal?, end: BigDecimal?): Boolean {
        start ?: return false
        end ?: return false
        return abs(end.subtract(start).toDouble() - 18_000.0) < 1.0
    }

    private fun exactLong(value: BigDecimal): Long? = try {
        value.longValueExact()
    } catch (_: ArithmeticException) {
        null
    }

    private fun decimal(entry: JsonElement, key: String): BigDecimal? =
        DecimalValue.from((entry as? JsonObject)?.get(key))?.value

    private fun string(entry: JsonElement, key: String): String? =
        ((entry as? JsonObject)?.get(key) as? JsonPrimitive)?.takeIf { it.isString }?.content
}
